Packages/MIES/MIES_ThreadsafeUtilities.ipf

```diff
@@ -165,11 +165,18 @@ threadsafe Function TS_ThreadGroupPutDFR(tgID, dfr)
 	DFREF dfr
 
 	string dataFolder
+	DFREF dfrSave
 
 	ASSERT_TS(DataFolderExistsDFR(dfr), "ThreadGroupPutDFR: dfr does not exist")
 
-	SetDataFolder dfr
-	ThreadGroupPutDF tgId, :
+	dataFolder = UniqueDataFolderName($":", "temp")
+	DuplicateDataFolder dfr, $dataFolder
+	dfrSave = GetDataFolderDFR()
+
+	SetDataFolder $dataFolder
+	ThreadGroupPutDF tgID, :
+
+	SetDatafolder dfrSave
```

Comment:
@t-b the next time we chat (no rush) may we walk through the rationale some of the see lines of code? - it's for my own edification.

Reply:
@timjarsky Sure!

Reply:
@t-b @timjarsky I am sorry, but I had to fix another flaw. So the code has changed slightly.

The main change is that before
`	dataFolder = UniqueDataFolderName($":", "temp")
`
a
`	SetDataFolder root:
`
is added.

If we don't do that and the current datafolder is a free data folder when the function is called, then UniqueDataFolderName returns "<number>:temp", where number is the free data folder id. However such DF name is not usable by `DuplicateDataFolder`. (would need just ":temp")
If DuplicateDataFolder fails here then the function crashes Igor. This is because then the `SetDataFolder `in front of `ThreadGroupPutDF `fails as well. So `ThreadGroupPutDF `operates on the current DF and not on the duplicated. As there are usually some references in use, Igor crashes.

So I added a try catch if DuplicateDataFolder fails, to stop execution in case before Igor crashes.

Secondary consideration:

UniqueDataFolderName should probably ASSERT if dfr is a free data folder as the returned DF path is not usable then.

What do you think?

Reply:
> UniqueDataFolderName should probably ASSERT if dfr is a free data folder as the returned 
> DF path is not usable then.

Yes let it assert out. I don't think it makes sense to be used with a free datafolder.

Reply:
> SetDataFolder root: 

Can we only do that in the case we need to? Our general approach, which worked really well, is that we only use `SetDataFolder` if there is no other way. IIRC `DataFolderRefStatus` can tell you if the current datafolder is a free data folder.

Reply:
@t-b From you comment in the chat recently I thought about a way to use generally a free datafolder to duplicate into, which would let us get rid of UniqueDataFolderName in this code part.

So I reworked the whole function.

Maybe there is even a faster way to find out if dfr is the root of a free DF?

Reply:
Looks good!

Reply:
@MichaelHuth I looked into that again with Tim and how about using `GetIndexedObjName` instead of manually deriving dfrName?

Reply:
You mean instead of `dfrName = "freeroot"` ?

Thats given by Igor and not exactly a manual derivation.

Reply:
@MichaelHuth yes!

Reply:
Ok, though conceptionally I think that such constants could be derived at MIES startup and reused.

Reply:
The whole drfName derivation can be replaced by `GetIndexedObjName` or?

Reply:
Yes, just pushed that. It assumes however that the free to free datafodler duplicate still results in a new datafolder. It also uses the fact that for no DF found an empty string is returned.

Reply:
@MichaelHuth Nice!